piq/vsi.py

```diff
@@ -0,0 +1,285 @@
+    vs_target = avg_pool2d(vs_target.unsqueeze(1), kernel_size=kernel_size).squeeze(1)
+
+    prediction_LMN = avg_pool2d(prediction_LMN, kernel_size=kernel_size)
+    target_LMN = avg_pool2d(target_LMN, kernel_size=kernel_size)
+
+    gm_prediction = scharr_grad_map(prediction_LMN[:, 1].unsqueeze(1)).squeeze(1)
+    gm_target = scharr_grad_map(target_LMN[:, 1].unsqueeze(1)).squeeze(1)
+
+    s_vs = (2 * vs_prediction * vs_target + c1) / (vs_prediction.pow(2) + vs_target.pow(2) + c1)
+    s_gm = (2 * gm_prediction * gm_target + c2) / (gm_prediction.pow(2) + gm_target.pow(2) + c2)
+
+    s_m = (2 * prediction_LMN[:, 1] * target_LMN[:, 1] + c3) / (
+            prediction_LMN[:, 1].pow(2) + target_LMN[:, 1].pow(2) + c3)
+    s_n = (2 * prediction_LMN[:, 2] * target_LMN[:, 2] + c3) / (
+            prediction_LMN[:, 2].pow(2) + target_LMN[:, 2].pow(2) + c3)
+    s_c = s_m * s_n
+
+    s = s_vs * s_gm.pow(alpha) * s_c.relu().pow(beta)
+
+    vs_max = torch.max(vs_prediction, vs_target)
+
+    eps = torch.finfo(vs_max.dtype).eps
+    output = ((s * vs_max).sum(dim=(-1, -2)) + eps) / (vs_max.sum(dim=(-1, -2)) + eps)
+
+    if reduction == 'none':
+        return output
+    return {'mean': torch.mean,
+            'sum': torch.sum
+            }[reduction](output, dim=0)
+
+
+class VSILoss(_Loss):
+    def __init__(self, reduction: str = 'mean', c1: float = 1.27, c2: float = 386., c3: float = 130.,
+                 alpha: float = 0.4, beta: float = 0.02, data_range: float = 1.,
+                 omega_0: float = 0.021, sigma_f: float = 1.34, sigma_d: float = 145., sigma_c: float = 0.001) -> None:
+
+        r"""Creates a criterion that measures Visual Saliency-induced Index error between
+            each element in the input and target.
+
+            The sum operation still operates over all the elements, and divides by :math:`n`.
+
+            The division by :math:`n` can be avoided if one sets ``reduction = 'sum'``.
+
+            Args:
+                reduction: Reduction over samples in batch: "mean"|"sum"|"none"
+                data_range: Value range of input images (usually 1.0 or 255). Default: 1.0
+                c1: coefficient to calculate saliency component of VSI
+                c2: coefficient to calculate gradient component of VSI
+                c3: coefficient to calculate color component of VSI
+                alpha: power for gradient component of VSI
+                beta: power for color component of VSI
+                omega_0: coefficient to get log Gabor filter at SDSP
+                sigma_f: coefficient to get log Gabor filter at SDSP
+                sigma_d: coefficient to get SDSP
+                sigma_c: coefficient to get SDSP
+
+            Shape:
+                - Input: Required to be 2D (H, W), 3D (C,H,W), 4D (N,C,H,W), channels first.
+                - Target: Required to be 2D (H, W), 3D (C,H,W), 4D (N,C,H,W), channels first.
```

Comment:
This is a great way to describe input requirements. I am curios why the same is not present in the `vsi` function 🤔